File: apps/ios/Auth/Sources/IbiminaAuthKit/AuthFlow/AuthAPI.swift
import Foundation

struct StartWhatsAppResponse: Decodable {
    let attemptId: String
    let expiresIn: Int
    let resendIn: Int?
}

struct VerifyWhatsAppResponse: Decodable {
    let token: String
    let refreshToken: String?
    let expiresIn: Int?
}

protocol AuthNetworking {
    func start(number: String) async throws -> StartWhatsAppResponse
    func verify(number: String, code: String, attemptId: String?) async throws -> VerifyWhatsAppResponse
}

final class AuthAPI: AuthNetworking {
    private let session: URLSession
    private let baseURL: URL

    init(baseURL: URL = URL(string: "https://api.example.com")!, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    func start(number: String) async throws -> StartWhatsAppResponse {
        let body = ["whatsappNumber": number]
        let request = try makeRequest(path: "/auth/whatsapp/start", body: body)
        let (data, response) = try await session.data(for: request)
        try validate(response: response)
        return try JSONDecoder().decode(StartWhatsAppResponse.self, from: data)
    }

    func verify(number: String, code: String, attemptId: String?) async throws -> VerifyWhatsAppResponse {
        let body: [String: Any?] = [
            "whatsappNumber": number,
            "code": code,
            "attemptId": attemptId
        ]
        let request = try makeRequest(path: "/auth/whatsapp/verify", body: body)
        let (data, response) = try await session.data(for: request)
        try validate(response: response)
        return try JSONDecoder().decode(VerifyWhatsAppResponse.self, from: data)
    }

    private func makeRequest(path: String, body: [String: Any?]) throws -> URLRequest {
        guard let url = URL(string: path, relativeTo: baseURL) else {
            throw NSError(domain: "AuthAPI", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid URL path"])
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body.compactMapValues { $0 })
        return request
    }

    private func validate(response: URLResponse) throws {
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw NSError(domain: "AuthAPI", code: http.statusCode, userInfo: [NSLocalizedDescriptionKey: "Request failed with status code \(http.statusCode)"])
        }
    }
}

File: archive/apps/android-auth/app/src/main/java/com/ibimina/auth/api/AuthApi.kt
package com.ibimina.auth.api

import retrofit2.http.Body
import retrofit2.http.POST

private const val DEFAULT_BASE_URL = "https://api.example.com"

data class StartWhatsAppRequest(
    val whatsappNumber: String
)

data class StartWhatsAppResponse(
    val attemptId: String,
    val expiresIn: Int,
    val resendIn: Int? = null
)

data class VerifyWhatsAppRequest(
    val whatsappNumber: String,
    val code: String,
    val attemptId: String?
)

data class VerifyWhatsAppResponse(
    val token: String,
    val refreshToken: String? = null,
    val expiresIn: Int? = null
)

interface AuthService {
    @POST("/auth/whatsapp/start")
    suspend fun start(@Body body: StartWhatsAppRequest): StartWhatsAppResponse

    @POST("/auth/whatsapp/verify")
    suspend fun verify(@Body body: VerifyWhatsAppRequest): VerifyWhatsAppResponse
}

fun provideAuthService(baseUrl: String = DEFAULT_BASE_URL): AuthService {
    val moshi = com.squareup.moshi.Moshi.Builder().build()
    val retrofit = retrofit2.Retrofit.Builder()
        .baseUrl(baseUrl)
        .addConverterFactory(com.squareup.retrofit2.converter.moshi.MoshiConverterFactory.create(moshi))
        .client(
            okhttp3.OkHttpClient.Builder()
                .addInterceptor { chain ->
                    val request = chain.request().newBuilder()
                        .addHeader("Accept", "application/json")
                        .build()
                    chain.proceed(request)
                }
                .build()
        )
        .build()
    return retrofit.create(AuthService::class.java)
}
